### PR TITLE
Support for schema evolution, staging, selective column projection and compatibility check for Avro to ORC

### PR DESCRIPTION
**Pull request by abti**
Adding support for:
- Schema evolution: Support to evolve destination ORC table for changes in source table schema (Hive / ORC does not support evolution yet, this is being added with hope that Hive / ORC will soon support evolution)
- Staging: Support to first stage converted data in a staging table before publishing / committing it to the final table
- Selective column projection: To counter against schema evolution on source hive table, this feature would help selectively project columns from source that are in destination / ORC table

### CODE REVIEW

**Comment by abti** [comment]
@pcadabam Can you please review?

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/publisher/HiveConvertPublisher.java [hunk 1]
```diff
@@ -57,13 +67,43 @@ public void initialize() throws IOException {
...
+        }
         wus.setWorkingState(WorkingState.FAILED);
         new SlaEventSubmitter(eventSubmitter, EventConstants.CONVERSION_FAILED_EVENT, wus.getProperties()).submit();
       }
     } else {
+      String publishTableCommands = StringUtils.EMPTY;
       for (WorkUnitState wus : states) {
+        if (isFirst) {
+          // Get publish table commands
+          publishTableCommands = HiveAvroORCQueryUtils.GSON
```
Comment (pcadabam) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/publisher/HiveConvertPublisher.java]: I think reading and writing the publish commands should be hidden inside ``` HiveAvroORCQueryUtils```. The publisher and converter should not know the workunit key or the serialization semantics.

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/dataset/ConvertibleHiveDataset.java [hunk 2]
```diff
@@ -161,7 +163,9 @@ private ConversionConfig(Config config, Table table, String destinationFormat) {
       // Optional
       this.clusterBy = ConfigUtils.getStringList(config, CLUSTER_BY_KEY);
       this.numBuckets = Optional.fromNullable(ConfigUtils.getInt(config, NUM_BUCKETS_KEY, null));
-      this.hiveRuntimeProperties = ConfigUtils.configToProperties(ConfigUtils.getConfig(config, HIVE_RUNTIME_PROPERTIES_KEY_PREFIX, ConfigFactory.empty()));
+      this.hiveRuntimeProperties = ConfigUtils
+          .configToProperties(ConfigUtils.getConfig(config, HIVE_RUNTIME_PROPERTIES_KEY_PREFIX, ConfigFactory.empty()));
+      this.evolutionEnabled = config.hasPath(EVOLUTION_ENABLED) && config.getBoolean(EVOLUTION_ENABLED);
```
Comment (pcadabam) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/dataset/ConvertibleHiveDataset.java]: Use ```ConfigUtils.getBoolean(Config config, String path, boolean def)```

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/converter/AbstractAvroToOrcConverter.java [hunk 3]
```diff
@@ -119,8 +127,11 @@ public String getConfigPrefix() {
 
     // ORC table name and location
     String orcTableName = getConversionConfig().getDestinationTableName();
+    String orcStagingTableName = orcTableName + "_staging"; // Fixed and non-configurable
```
Comment (pcadabam) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/converter/AbstractAvroToOrcConverter.java]: I think this should go inside ConversionConfig

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/util/HiveAvroORCQueryUtils.java [hunk 4]
```diff
@@ -35,6 +41,11 @@
 @Slf4j
 public class HiveAvroORCQueryUtils {
 
+  public static final String SERIALIZED_PUBLISH_TABLE_COMMANDS = "serialized.publish.table.commands";
```
Comment (pcadabam) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/util/HiveAvroORCQueryUtils.java]: These should be ```private```

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/publisher/HiveConvertPublisher.java (hunk 1 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/publisher/HiveConvertPublisher.java]: Moved

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/dataset/ConvertibleHiveDataset.java (hunk 2 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/dataset/ConvertibleHiveDataset.java]: Changed

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/converter/AbstractAvroToOrcConverter.java (hunk 3 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/converter/AbstractAvroToOrcConverter.java]: Moved

File: gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/util/HiveAvroORCQueryUtils.java (hunk 4 above)
Comment (abti) [inline gobblin-data-management/src/main/java/gobblin/data/management/conversion/hive/util/HiveAvroORCQueryUtils.java]: Changed

**Comment by pcadabam** [comment]
LGTM

**Pull request merged**